File: frontend/src/api/simulationHistory.js
// frontend/src/api/simulationHistory.js
//
// 설정 저장·Replay·시점 복원 관련 API 클라이언트.
// 기준: API_명세_및_공통_규약.md v1.6, §3.7~3.13
// 기존 client.js(apiRequest) 컨벤션을 그대로 따른다:
//   - apiRequest(path, { token, method, body, headers })
//   - 성공 시 응답 JSON({data} 또는 {data, meta})을 그대로 반환 → 이 파일에서 .data를 꺼내 돌려줌
//   - 실패 시 이미 status/code/message가 채워진 Error를 throw (별도 에러 클래스 불필요)

import { apiRequest } from "./client.js";

// ---------------------------------------------------------------------------
// 3.11 설정 저장·변경
// ---------------------------------------------------------------------------

/**
 * Draft 상태 Simulation의 초기 파라미터 전체 저장/변경.
 * PUT /simulations/{simulation_id}/parameters
 */
export async function saveDraftConfig(token, simulationId, config) {
  const result = await apiRequest(`/v1/simulations/${simulationId}/parameters`, {
    token,
    method: "PUT",
    body: JSON.stringify(config),
  });
  return result.data ?? result;
}

/**
 * 실행 중(RUNNING/PAUSED) Simulation의 event_frequency, event_impact만 변경.
 * PATCH /simulations/{simulation_id}/parameters
 */
export async function updateRunningConfig(token, simulationId, { eventFrequency, eventImpact }) {
  const result = await apiRequest(`/v1/simulations/${simulationId}/parameters`, {
    token,
    method: "PATCH",
    body: JSON.stringify({
      event_frequency: eventFrequency,
      event_impact: eventImpact,
    }),
  });
  return result.data ?? result;
}

// ---------------------------------------------------------------------------
// 3.9 / 3.10 시점 조회 · 복원
// ---------------------------------------------------------------------------

/**
 * 특정 Tick 시점 상태 조회 (읽기 전용, 부수 효과 없음, 새 Simulation 생성 안 함).
 * GET /simulations/{simulation_id}/snapshots/{tick_number}
 */
export async function getSnapshot(token, simulationId, tickNumber) {
  const result = await apiRequest(
    `/v1/simulations/${simulationId}/snapshots/${tickNumber}`,
    { token }
  );
  return result.data ?? result;
}

/**
 * 저장본 복원.
 *
 * ⚠️ 백엔드 구현(SimulationSnapshotService.restore_as_branch) 기준:
 * 원본 Simulation을 제자리에서 갱신하지 않고, origin_simulation_id /
 * origin_snapshot_id로 원본을 참조하는 새 Simulation을 생성해 반환한다.
 * 호출 측(컴포넌트)에서는 응답의 id를 원본과 동일하다고 가정하지 말고,
 * 반환된 새 id로 이동(navigate)해야 한다.
 *
 * POST /simulations/{simulation_id}/restore
 */
export async function restoreSnapshot(token, simulationId, snapshotId) {
  const result = await apiRequest(`/v1/simulations/${simulationId}/restore`, {
    token,
    method: "POST",
    body: JSON.stringify({ snapshot_id: snapshotId }),
  });
  return result.data ?? result;
}

// ---------------------------------------------------------------------------
// 3.12 / 3.13 Replay
// ---------------------------------------------------------------------------

/**
 * 재생 가능한 Tick 목록 조회.
 * GET /simulations/{simulation_id}/replay
 */
export async function getReplayList(token, simulationId, { cursor, limit = 20 } = {}) {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  if (cursor) params.set("cursor", cursor);

  const result = await apiRequest(
    `/v1/simulations/${simulationId}/replay?${params.toString()}`,
    { token }
  );
  return { items: result.data ?? [], meta: result.meta };
}

/**
 * 특정 Tick의 Replay 데이터(Event, Agent Snapshot, Relationship Delta) 조회.
 * GET /simulations/{simulation_id}/replay/{tick_number}
 */
export async function getReplayTick(token, simulationId, tickNumber) {
  const result = await apiRequest(
    `/v1/simulations/${simulationId}/replay/${tickNumber}`,
    { token }
  );
  return result.data ?? result;
}

File: backend/app/api/simulation_history.py
from collections.abc import Callable
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from fastapi.routing import APIRoute
from sqlalchemy.orm import Session

from app.api.schemas import (
    RestoreSnapshotRequest,
    SimulationConfigPatchRequest,
    SimulationConfigPutRequest,
)
from app.core.database import get_db
from app.core.security import require_user_role
from app.domain.models import Simulation, User
from app.repositories.simulation_snapshots import SimulationSnapshotRepository
from app.services.simulation_replay import (
    ReplayResourceNotFoundError,
    SimulationReplayService,
    SnapshotMismatchError,
    UnsupportedReplaySnapshotSchemaError,
)
from app.services.simulation_snapshots import (
    InitialSettingsLockedError,
    InvalidSimulationConfigError,
    SimulationConfigInput,
    SimulationSettingsLockedError,
    SimulationSnapshotService,
    UnsupportedSnapshotSchemaError,
)


class Slice6APIError(Exception):
    def __init__(self, status_code: int, code: str, message: str) -> None:
        self.status_code = status_code
        self.code = code
        self.message = message


def _error_response(status_code: int, code: str, message: str) -> JSONResponse:
    return JSONResponse(
        status_code=status_code,
        content={"error": {"code": code, "message": message}},
    )


class Slice6ErrorRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_handler = super().get_route_handler()

        async def handler(request: Request):
            try:
                return await original_handler(request)
            except Slice6APIError as exc:
                return _error_response(exc.status_code, exc.code, exc.message)
            except RequestValidationError:
                return _error_response(400, "INVALID_REPLAY_REQUEST", "Invalid request")
            except HTTPException as exc:
                codes = {
                    401: "AUTHENTICATION_REQUIRED",
                    403: "SIMULATION_ACCESS_DENIED",
                    404: "REPLAY_RESOURCE_NOT_FOUND",
                }
                return _error_response(
                    exc.status_code,
                    codes.get(exc.status_code, "INVALID_REPLAY_REQUEST"),
                    str(exc.detail),
                )

        return handler


router = APIRouter(
    prefix="/simulations",
    tags=["simulation-history"],
    route_class=Slice6ErrorRoute,
)
snapshot_service = SimulationSnapshotService()
replay_service = SimulationReplayService()
snapshot_repository = SimulationSnapshotRepository()


def _owned_simulation(db: Session, simulation_id: UUID, user: User) -> Simulation:
    simulation = db.get(Simulation, simulation_id)
    if simulation is None:
        raise Slice6APIError(404, "REPLAY_RESOURCE_NOT_FOUND", "Simulation not found")
    if simulation.owner_id != user.id:
        raise Slice6APIError(403, "SIMULATION_ACCESS_DENIED", "Simulation access denied")
    return simulation


def _config_data(config) -> dict[str, Any]:
    return {
        "event_frequency": config.event_frequency,
        "event_impact": config.event_impact,
        "magic_enabled": config.magic_enabled,
        "config_version": config.version,
        "changed_at": config.created_at.isoformat(),
    }


def _save_config(
    db: Session,
    simulation: Simulation,
    config_input: SimulationConfigInput,
) -> JSONResponse:
    try:
        config = snapshot_service.save_config(db, simulation, config_input)
        db.commit()
        db.refresh(config)
    except SimulationSettingsLockedError as exc:
        db.rollback()
        raise Slice6APIError(409, "SIMULATION_SETTINGS_LOCKED", str(exc)) from exc
    except InitialSettingsLockedError as exc:
        db.rollback()
        raise Slice6APIError(409, "INITIAL_SETTINGS_LOCKED", str(exc)) from exc
    except InvalidSimulationConfigError as exc:
        db.rollback()
        raise Slice6APIError(400, "INVALID_REPLAY_REQUEST", str(exc)) from exc
    return JSONResponse(content={"data": _config_data(config)})


@router.put("/{simulation_id}/parameters")
def put_parameters(
    simulation_id: UUID,
    request: SimulationConfigPutRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user_role),
) -> JSONResponse:
    simulation = _owned_simulation(db, simulation_id, current_user)
    latest = snapshot_service.configs.latest(db, simulation.id)
    return _save_config(
        db,
        simulation,
        SimulationConfigInput(
            event_frequency=request.event_frequency,
            event_impact=request.event_impact,
            magic_enabled=request.magic_enabled,
            user_persona_settings=(latest.user_persona_settings if latest else {}),
            policy_version=(latest.policy_version if latest else None),
            resolver_version=(latest.resolver_version if latest else None),
        ),
    )


@router.patch("/{simulation_id}/parameters")
def patch_parameters(
    simulation_id: UUID,
    request: SimulationConfigPatchRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user_role),
) -> JSONResponse:
    simulation = _owned_simulation(db, simulation_id, current_user)
    latest = snapshot_service.configs.latest(db, simulation.id)
    if latest is None:
        raise Slice6APIError(409, "INITIAL_SETTINGS_LOCKED", "Initial settings not found")
    return _save_config(
        db,
        simulation,
        SimulationConfigInput(
            event_frequency=request.event_frequency,
            event_impact=request.event_impact,
            magic_enabled=latest.magic_enabled,
            user_persona_settings=latest.user_persona_settings,
            policy_version=latest.policy_version,
            resolver_version=latest.resolver_version,
        ),
    )


@router.get("/{simulation_id}/snapshots/{tick_number}")
def get_snapshot(
    simulation_id: UUID,
    tick_number: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user_role),
) -> dict[str, Any]:
    _owned_simulation(db, simulation_id, current_user)
    snapshot = snapshot_repository.get_at_tick(db, simulation_id, tick_number)
    if snapshot is None or snapshot.simulation_id != simulation_id:
        raise Slice6APIError(404, "REPLAY_RESOURCE_NOT_FOUND", "Snapshot not found")
    payload = dict(snapshot.payload)
    return {
        "data": {
            "snapshot_id": str(snapshot.id),
            "tick_number": snapshot.tick_number,
            "simulation_day": payload.get("simulation", {}).get("current_day"),
            **payload,
        }
    }


@router.post("/{simulation_id}/restore")
def restore_snapshot(
    simulation_id: UUID,
    request: RestoreSnapshotRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user_role),
) -> dict[str, Any]:
    _owned_simulation(db, simulation_id, current_user)
    snapshot = snapshot_repository.get(db, request.snapshot_id)
    if snapshot is None:
        raise Slice6APIError(404, "REPLAY_RESOURCE_NOT_FOUND", "Snapshot not found")
    source = db.get(Simulation, snapshot.simulation_id)
    if source is None:
        raise Slice6APIError(404, "REPLAY_RESOURCE_NOT_FOUND", "Simulation not found")
    if source.owner_id != current_user.id:
        raise Slice6APIError(403, "SIMULATION_ACCESS_DENIED", "Simulation access denied")
    if snapshot.simulation_id != simulation_id:
        raise Slice6APIError(409, "SNAPSHOT_MISMATCH", "Snapshot does not match simulation")
    try:
        payload = snapshot_service.restore_snapshot(
            db, snapshot.id, owner_id=current_user.id
        )
    except UnsupportedSnapshotSchemaError as exc:
        raise Slice6APIError(409, "UNSUPPORTED_SNAPSHOT_SCHEMA", str(exc)) from exc
    return {"data": payload}


@router.get("/{simulation_id}/replay")
def get_replay_list(
    simulation_id: UUID,
    limit: int = Query(default=20, ge=1, le=100),
    cursor: int | None = Query(default=None, ge=0),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user_role),
) -> dict[str, Any]:
    _owned_simulation(db, simulation_id, current_user)
    items, has_more = replay_service.list_ticks(
        db, simulation_id, after_tick=cursor, limit=limit
    )
    next_cursor = str(items[-1]["tick_number"]) if has_more and items else None
    return {
        "data": items,
        "meta": {"next_cursor": next_cursor, "has_more": has_more},
    }


@router.get("/{simulation_id}/replay/{tick_number}")
def get_replay_tick(
    simulation_id: UUID,
    tick_number: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_user_role),
) -> dict[str, Any]:
    _owned_simulation(db, simulation_id, current_user)
    try:
        return {"data": replay_service.get_tick(db, simulation_id, tick_number)}
    except ReplayResourceNotFoundError as exc:
        raise Slice6APIError(404, "REPLAY_RESOURCE_NOT_FOUND", str(exc)) from exc
    except SnapshotMismatchError as exc:
        raise Slice6APIError(409, "SNAPSHOT_MISMATCH", str(exc)) from exc
    except UnsupportedReplaySnapshotSchemaError as exc:
        raise Slice6APIError(409, "UNSUPPORTED_SNAPSHOT_SCHEMA", str(exc)) from exc
